### PR TITLE
feat: Customize workfow metadata from event data

### PR DESCRIPTION
**Pull request by michaeljohnalbers**
Signed-off-by: Michael Albers <michael.john.albers@gmail.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Fixes https://github.com/argoproj/argo/issues/4640

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/event_types.go [hunk 1]
```diff
@@ -40,6 +40,17 @@ type Submit struct {
 	// WorkflowTemplateRef the workflow template to submit
 	WorkflowTemplateRef WorkflowTemplateRef `json:"workflowTemplateRef" protobuf:"bytes,1,opt,name=workflowTemplateRef"`
 
+	// WorkflowName optional means to customize the name of the submitted workflow
+	WorkflowName *WorkflowName `json:"workflowName,omitempty" protobuf:"bytes,2,opt,name=workflowName"`
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/event_types.go:44]: this is great. 

* We could allow for future changes, e.g. templating labels and annotations.
* We could make the names fully based on the expression environment - no suffix. 

```yaml
metadata:
  name: payload.message + timestamp
```

Useful as follows:

```yaml
metadata:
  name: payload.github.repo + "-" + payload.github.sha1
```

Maybe in the future:

```yaml
metadata:
  annotations: 
    sha1: payload.github.sha1
```

File: pkg/apis/workflow/v1alpha1/event_types.go [hunk 2]
```diff
@@ -40,6 +40,17 @@ type Submit struct {
...
-	Arguments *Arguments `json:"arguments,omitempty" protobuf:"bytes,2,opt,name=arguments"`
+	Arguments *Arguments `json:"arguments,omitempty" protobuf:"bytes,3,opt,name=arguments"`
+}
+
+type WorkflowName struct {
+	// Value is the literal value to use for the workflow name.
+	Value *AnyString `json:"value,omitempty" protobuf:"bytes,1,opt,name=value"`
+
+	// ValueFrom is the source for the workflow name
+	ValueFrom *ValueFrom `json:"valueFrom,omitempty" protobuf:"bytes,2,opt,name=valueFrom"`
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/event_types.go:55]: I don't think you need `value` or `valueFrom` at all

File: server/event/dispatch/operation.go [hunk 3]
```diff
@@ -129,6 +137,31 @@ func (o *Operation) dispatch(wfeb wfv1.WorkflowEventBinding, nameSuffix string)
...
+		name = wfn.Value.String()
+	} else if wfn.ValueFrom != nil {
+		result, err := expr.Eval(wfn.ValueFrom.Event, o.env)
+		if err != nil {
+			return nil, fmt.Errorf("failed to evaluate workflow name expression: %w", err)
+		}
+
+		// Prevent expressions that evaluate to types which will generate invalid names.
+		switch v := result.(type) {
+		case string:
```
Comment (alexec) [inline server/event/dispatch/operation.go:152]: can just `return  v, nil` here

**Comment by michaeljohnalbers** [comment]
@alexec The codegen check is failing. The comment in the PR creation form says to run "make codegen" to fix it. When I run that nothing changes in the git repo. Any ideas on what I can try?

File: pkg/apis/workflow/v1alpha1/event_types.go (hunk 2 above)
Comment (michaeljohnalbers) [reply pkg/apis/workflow/v1alpha1/event_types.go:55]: What should I be using instead? I used those to stay somewhat similar to how parameters are defined.

**Comment by alexec** [comment]
Option 1 

* You can force files to be updated by deleting the change files and running again.
* What I do: copy the diff from the build job and run `pbpaste|cut -c 30-|git apply`

File: pkg/apis/workflow/v1alpha1/event_types.go (hunk 2 above)
Comment (alexec) [reply pkg/apis/workflow/v1alpha1/event_types.go:55]: my comment is irrelevant - I want you to replace `workflowName` with `metadata`

**Comment by michaeljohnalbers** [comment]
@alexec Codegen is still failing. It's complaining about the changes I made to the operation_test.go file. Why is that? I assume you'd want unit tests added for this change.

**Comment by alexec** [comment]
You need to apply this patch to fix the codegen issue:

```
diff --git a/server/event/dispatch/operation_test.go b/server/event/dispatch/operation_test.go
index 1c9703ab..f75bb41d 100644
--- a/server/event/dispatch/operation_test.go
+++ b/server/event/dispatch/operation_test.go
@@ -199,8 +199,8 @@ func Test_populateWorkflowMetadata(t *testing.T) {
 				Submit: &wfv1.Submit{
 					WorkflowTemplateRef: wfv1.WorkflowTemplateRef{Name: "my-wft"},
 					ObjectMeta: metav1.ObjectMeta{
-						Name: "\"my-wfeb-2\"",
-						Labels: map[string]string{"aLabel": "\"someValue\""},
+						Name:        "\"my-wfeb-2\"",
+						Labels:      map[string]string{"aLabel": "\"someValue\""},
 						Annotations: map[string]string{"anAnnotation": "\"otherValue\""},
 					},
 				},
@@ -214,8 +214,8 @@ func Test_populateWorkflowMetadata(t *testing.T) {
 				Submit: &wfv1.Submit{
 					WorkflowTemplateRef: wfv1.WorkflowTemplateRef{Name: "my-wft"},
 					ObjectMeta: metav1.ObjectMeta{
-						Name: "\"my-wfeb-\" + payload.foo.bar",
-						Labels: map[string]string{"aLabel": "payload.list[0]"},
+						Name:        "\"my-wfeb-\" + payload.foo.bar",
+						Labels:      map[string]string{"aLabel": "payload.list[0]"},
 						Annotations: map[string]string{"anAnnotation": "payload.list[1]"},
 					},
 				},
@@ -228,7 +228,7 @@ func Test_populateWorkflowMetadata(t *testing.T) {
 				Event: wfv1.Event{Selector: "true"},
 				Submit: &wfv1.Submit{
 					WorkflowTemplateRef: wfv1.WorkflowTemplateRef{Name: "my-wft"},
-					ObjectMeta: metav1.ObjectMeta{Name: "payload.......foo[.numeric]"},
+					ObjectMeta:          metav1.ObjectMeta{Name: "payload.......foo[.numeric]"},
 				},
 			},
 		},
@@ -263,7 +263,7 @@ func Test_populateWorkflowMetadata(t *testing.T) {
 				Event: wfv1.Event{Selector: "true"},
 				Submit: &wfv1.Submit{
 					WorkflowTemplateRef: wfv1.WorkflowTemplateRef{Name: "my-wft"},
-					ObjectMeta: metav1.ObjectMeta{Name: "payload.foo.numeric"},
+					ObjectMeta:          metav1.ObjectMeta{Name: "payload.foo.numeric"},
 				},
 			},
 		},
@@ -274,7 +274,7 @@ func Test_populateWorkflowMetadata(t *testing.T) {
 				Event: wfv1.Event{Selector: "true"},
 				Submit: &wfv1.Submit{
 					WorkflowTemplateRef: wfv1.WorkflowTemplateRef{Name: "my-wft"},
-					ObjectMeta: metav1.ObjectMeta{Name: "payload.foo.bool"},
+					ObjectMeta:          metav1.ObjectMeta{Name: "payload.foo.bool"},
 				},
 			},
 		},
@@ -285,7 +285,7 @@ func Test_populateWorkflowMetadata(t *testing.T) {
 				Event: wfv1.Event{Selector: "true"},
 				Submit: &wfv1.Submit{
 					WorkflowTemplateRef: wfv1.WorkflowTemplateRef{Name: "my-wft"},
-					ObjectMeta: metav1.ObjectMeta{Name: "payload.foo"},
+					ObjectMeta:          metav1.ObjectMeta{Name: "payload.foo"},
 				},
 			},
 		},
@@ -296,7 +296,7 @@ func Test_populateWorkflowMetadata(t *testing.T) {
 				Event: wfv1.Event{Selector: "true"},
 				Submit: &wfv1.Submit{
 					WorkflowTemplateRef: wfv1.WorkflowTemplateRef{Name: "my-wft"},
-					ObjectMeta: metav1.ObjectMeta{Name: "payload.list"},
+					ObjectMeta:          metav1.ObjectMeta{Name: "payload.list"},
 				},
 			},
 		},
@@ -307,12 +307,12 @@ func Test_populateWorkflowMetadata(t *testing.T) {
 				Event: wfv1.Event{Selector: "true"},
 				Submit: &wfv1.Submit{
 					WorkflowTemplateRef: wfv1.WorkflowTemplateRef{Name: "my-wft"},
-					ObjectMeta: metav1.ObjectMeta{Name: "payload.nothing"},
+					ObjectMeta:          metav1.ObjectMeta{Name: "payload.nothing"},
 				},
 			},
 		},
 	}, "my-ns", "my-discriminator",
-	&wfv1.Item{Value: json.RawMessage(`{"foo": {"bar": "baz", "numeric": 8675309, "bool": true}, "list": ["one", "two"]}`)})
+		&wfv1.Item{Value: json.RawMessage(`{"foo": {"bar": "baz", "numeric": 8675309, "bool": true}, "list": ["one", "two"]}`)})
 
 	assert.NoError(t, err)
 	operation.Dispatch()
```

File: pkg/apis/workflow/v1alpha1/event_types.go [hunk 4]
```diff
@@ -40,6 +40,9 @@ type Submit struct {
 	// WorkflowTemplateRef the workflow template to submit
 	WorkflowTemplateRef WorkflowTemplateRef `json:"workflowTemplateRef" protobuf:"bytes,1,opt,name=workflowTemplateRef"`
 
+	// Metadata optional means to customize select fields of the workflow metadata
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,2,opt,name=metadata"`
+
 	// Arguments extracted from the event and then set as arguments to the workflow created.
-	Arguments *Arguments `json:"arguments,omitempty" protobuf:"bytes,2,opt,name=arguments"`
+	Arguments *Arguments `json:"arguments,omitempty" protobuf:"bytes,3,opt,name=arguments"`
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/event_types.go:47]: it is important not to change these numbers as it breaks clients - can you leave this as 3 and make metadata 2?

File: server/event/dispatch/operation.go [hunk 5]
```diff
@@ -98,8 +98,16 @@ func (o *Operation) dispatch(wfeb wfv1.WorkflowEventBinding, nameSuffix string)
...
 		wf := common.NewWorkflowFromWorkflowTemplate(tmpl.GetName(), tmpl.GetWorkflowMetadata(), ref.ClusterScope)
 		o.instanceIDService.Label(wf)
-		// make sure we have a predicable name, so re-creation doesn't create two workflows
-		wf.SetName(wf.GetGenerateName() + nameSuffix)
+		err = o.populateWorkflowMetadata(wf, &submit.ObjectMeta)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(wf.Name) == 0 {
```
Comment (alexec) [inline server/event/dispatch/operation.go:106]: minor - I'd prefer `wf.Name == ""`

File: server/event/dispatch/operation.go [hunk 6]
```diff
@@ -129,6 +137,45 @@ func (o *Operation) dispatch(wfeb wfv1.WorkflowEventBinding, nameSuffix string)
...
+			return err
+		}
+		wf.SetName(evalName)
+	}
+	for labelKey, labelValue := range metadata.Labels {
+		evalLabel, err := o.evaluateStringExpression(labelValue, fmt.Sprintf("label \"%s\"", labelKey))
+		if err != nil {
+			return err
+		}
+		wf.Labels[labelKey] = evalLabel
```
Comment (alexec) [inline server/event/dispatch/operation.go:153]: I think you might need to check if `wf.Labels == nil` here

File: server/event/dispatch/operation.go [hunk 7]
```diff
@@ -129,6 +137,45 @@ func (o *Operation) dispatch(wfeb wfv1.WorkflowEventBinding, nameSuffix string)
...
+	return nil
+}
+
+func (o *Operation) evaluateStringExpression(statement string, errorInfo string) (string, error) {
+	result, err := expr.Eval(statement, o.env)
+	if err != nil {
+		return "", fmt.Errorf("failed to evaluate workflow %s expression: %w", errorInfo, err)
+	}
+
+	switch v := result.(type) {
```
Comment (alexec) [inline server/event/dispatch/operation.go:171]: minor - `v, ok := result.(string)` is a more conventional way of doing this

**Review by alexec — CHANGES_REQUESTED** [review]
Looks great - thank you for going above-and-beyond with also implementing labels and annotations - I wasn't expecting this!

File: server/event/dispatch/operation.go (hunk 7 above)
Comment (michaeljohnalbers) [reply server/event/dispatch/operation.go]: Fixed. Thanks for showing me this.

File: server/event/dispatch/operation.go (hunk 6 above)
Comment (michaeljohnalbers) [reply server/event/dispatch/operation.go:158]: Fixed. Did the same for annotations.

File: server/event/dispatch/operation.go (hunk 5 above)
Comment (michaeljohnalbers) [reply server/event/dispatch/operation.go]: Fixed.

File: pkg/apis/workflow/v1alpha1/event_types.go (hunk 4 above)
Comment (michaeljohnalbers) [reply pkg/apis/workflow/v1alpha1/event_types.go]: I knew I shouldn't have done that. Fixed.

**Review by alexec — APPROVED** [review]
LGTM

**Comment by paveq** [comment]
I wonder if this supports getting name from metadata (HTTP headers), I tried following but got errors in argo-server:

```
  submit:
    metadata:
      name: 'metadata["x-event-id"] + "-test"'
```

**Pull request merged**